### ISSUE TITLE
'Unknown Mobile Command' error when checking for isDisplayed

### ISSUE THREAD

**Issue by algernon28** [serenity-bdd/serenity-core#3351]
### What happened?

I am experiencing weird issues when I run my mobile tests thru SauceLabs.
While my tests all run fine in local, both with emulator and real device, my tests break when running on Saucelabs.

The problem seems to be on Serenity side, which is instancing RemoteWebDriver instead of AndroidDriver or IOSDriver, therefore making it impossible for sauce to run any commands specific to mobile.

I reached out SauceLabs and after some back and forth this is the response:

"_I managed to run the customer Serenity BDD framework on my local Android.
I changed and ran the test on Sauce Emulator (not Sauce Real Device) - to check and compare local emulator execution to Sauce emulator execution._
 
_The command that I ran is:_
`actor.should(seeThat(LoginPage.isLoaded())); //
` 
This command calls:
```
public static Question<Boolean> isLoaded() {
    return Visibility.of(TITLE);
}
```
 
The TITLE is a locator on the first page:
```
public static final Target TITLE = Target.the("Login title")
//            .locatedForAndroid(By.xpath("//android.widget.TextView[@text='Inserisci le tue credenziali']"))            
.locatedForAndroid(By.xpath("//android.widget.TextView[@text='Insert your credentials']"))
            .locatedForIOS(By.id("Inserisci le tue credenziali"));"
```
_The command "Visibility"   is a Serenity BDD command that checks if the element is visible.
On the local emulator, it runs as expected:
Find the "Title" locator
Calls AppiumDriver.elementDisplayed() with the locator id_
 
_But on the Sauce emulator, the Visibility command performs:
Find the "Title" locator ( as expected)
Execute a sync command that tries to inject a JS code of "isDisplayed" that can be done on web elements -> This command returns an error "NotImplementedError: Method is not implemented"
Checking for Alert window. When the element is not displayed, Appium checks if there is an alert that blocks this element
Appium calls to the getUrl method.  It seems that Serenity BDD thinks that this is a web app or a webview and calls this method that returns an error "Method has not yet been implemented" .
From this behaviour, we can understand that Serenity starts the Appium Session on Sauce devices as a web app or not in the native-app context ( the first page also has webView context, as can seen in the attachment)._
 
_I could only find this doc from Serenity about the integration between Sauce and Serenity:
https://serenity-bdd.github.io/docs/cloud/saucelabs
But this is for the Selenium framework.  There is no available documentation about the Appium integration._

_Reach out to Serenity on how to start a driver when running a native app test on Sauce Labs Emulators or real devices_"

[serenity.conf.zip](https://github.com/serenity-bdd/serenity-core/files/13621804/serenity.conf.zip)
[android-physical.properties.zip](https://github.com/serenity-bdd/serenity-core/files/13621809/android-physical.properties.zip)

### What did you expect to happen?

I expect the mobile tests run on cloud like they do in local.

### Serenity BDD version

4.0.24

### JDK version

17

### Execution environment

Windows 10 pro
Appium

### How to reproduce the bug.

Just run any mobile specific command through Saucelabs

**Comment by algernon28** [comment]
What I can see is that using my local device and Appium, the driver is an instance of AndroidDriver, while via Sauce is a RemoteWebDriver, and it seems it's looking for a web application. In my serenity.conf I tried both webdriver.driver= appium and webdriver.driver = remote, with the same result

**Comment by wakaleo** [comment]
Can you connect to SauceLabs other than with the RemoteWebDriver? That is the usual way to interact with a remote grid, but if there is another way, maybe you can investigate and propose a PR to incorporate this (such as adapting the appium driver creation logic). I am not very familiar with the Appium code (it is largely community driven) so would not be well placed to know what would work best.

**Comment by algernon28** [comment]
The issue is that with RemoteDriver SauceLabs (but I suspect other cloud providers as well) cannot determine that it's a mobile test. 
Therefore, it's impossible to perform any mobile specific commands, so it messes up the test because many selenium commands (like isDisplayed) are also not available on mobile.
The solution would be to instantiate **AndroidDriver** or **IOSDriver** instead of RemoteWebDriver for tests where webdriver.remote=appium.
I do not have enough Serenity knowledge to propose a PR, but I would think it's not something huge.
This problem basically impedes any mobile test thru Saucelabs with Serenity.

As they stated (the quoted text is from Saucelabs, who investigated on the problem):


_The command "Visibility" is a Serenity BDD command that checks if the element is visible.
On the local emulator, it runs as expected:
Find the "Title" locator
Calls AppiumDriver.elementDisplayed() with the locator id

But on the Sauce emulator, the Visibility command performs:
Find the "Title" locator ( as expected)
Execute a sync command that tries to inject a JS code of "isDisplayed" that can be done on web elements -> This command returns an error "NotImplementedError: Method is not implemented"
Checking for Alert window. When the element is not displayed, Appium checks if there is an alert that blocks this element
Appium calls to the getUrl method. It seems that Serenity BDD thinks that this is a web app or a webview and calls this method that returns an error "Method has not yet been implemented" .
From this behaviour, we can understand that Serenity starts the Appium Session on Sauce devices as a web app or not in the native-app context ( the first page also has webView context, as can seen in the attachment)._

**Comment by wakaleo** [comment]
Can you check out the latest code from master and do a local build and try with that? You will need to set:
```
webdriver.remote.url=appium
```

**Comment by algernon28** [comment]
Hi John, thanks for the reply, I'm gonna try asap and let you know

**Comment by algernon28** [comment]
I tried to build 3.2.6-SNAPSHOT from master branch, therefore reverting everything to Serenity BDD 3x
I tried this configuration as last (but I tried every possible combination)
        driver = remote
        remote.driver = appium
            remote.url="https://#{saucelabs.username}:#{saucelabs.accessKey}@ondemand.eu-central-1.saucelabs.com:443/wd/hub".
Apparently now it's not even able to instantiate the driver:
Given User can see the login page     # com.edenred.qa.multiwallet.authentication.LoginSteps.userCanSeeTheLoginPage(net.serenitybdd.screenplay.Actor)
      net.thucydides.core.webdriver.ThucydidesConfigurationException: The configured driver 'Uninitialised WebDriverFacade' does not support Cross Platform Mobile targets
This is the bit of relevant code: 
    _@Given("{actor} can see the login page")
    public void userCanSeeTheLoginPage(Actor actor) {
        WaitUntil.the(LoginPage.TITLE, isVisible()).forNoMoreThan(Duration.ofSeconds(30));
        actor.attemptsTo(Ensure.that(LoginPage.TITLE).isDisplayed()); //
    }_

And this is the LoginPage:
_public class LoginPage extends MobilePage {

    public static final Target TITLE = Target.the("Login title")
            .locatedForAndroid(By.xpath("//android.widget.TextView[@text='Inserisci le tue credenziali']"))
            .locatedForIOS(By.id(""));

    public static final Target USERNAME = Target.the("Username")
            .locatedForAndroid(By.xpath("//android.widget.EditText[@resource-id='Username']"))
            .locatedForIOS(By.xpath(""));
    public static final Target PASSWORD = Target.the("Password")
            .locatedForAndroid(By.xpath("//android.widget.EditText[@resource-id='Password']"))
            .locatedForIOS(By.xpath(""));
    public static final Target BTN_ENTER = Target.the("Submit")
            .locatedForAndroid(By.xpath("//android.widget.Button[@resource-id='login']"))
            .locatedForIOS(By.id("Accedi"));
    public static final Target REGISTER = Target.the("Register")
            .locatedForAndroid(By.xpath("//android.widget.TextView[contains(@text,'Registrati']"))
            .locatedForIOS(By.name(""));
    public static final Target RECOVER_PASSWORD = Target.the("Recover password")
            .locatedForAndroid(By.xpath("//android.widget.TextView[starts-with(@text, 'Password ')]"))
            .locatedForIOS(By.name("Password dimenticata"));
    public static final Target INVALID_CREDENTIALS_ERROR = Target.the("Invalid credentials error")
            .locatedForAndroid(By.xpath("//android.view.View[@resource-id='invalid_credentials']"))
            .locatedForIOS(By.xpath("//XCUIElementTypeStaticText[@name='Le credenziali d’accesso non sono corrette']"));
    public static final Target CLOSE_ERROR_POPUP = Target.the("Close error popup")
            .locatedForAndroid(By.xpath("//android.widget.Button[@text='Close']"))
            .locatedForIOS(By.xpath("//XCUIElementTypeButton[@name='Close']"));

    public static final Target INVALID_FORMAT_ERROR = ByMobileTarget.the("Invalid format error")
            .locatedForAndroid(By.xpath("//android.webkit.WebView[@text='Edenred Connect']/android.view.View[1]"))
            .locatedForIOS(By.id(""));
    /**
     * @return if the title is visible, indicating the page is loaded
     */
    public static Question<Boolean> isLoaded() {
        return Visibility.of(TITLE);
    }

    public static Target error(Authenticate.ERROR_MODALITY modality){
        return (modality == Authenticate.ERROR_MODALITY.POPUP)? INVALID_CREDENTIALS_ERROR : INVALID_FORMAT_ERROR;}
}_

So in short: with version 4.0.29 I still get the error I described, while reverting to 3.0.26-SNAPSHOT from master I am not even able to instantiate the driver at all.
I attach here the jar I built, just in case I missed something.

[serenity-core-3.2.6-SNAPSHOT.jar.zip](https://github.com/serenity-bdd/serenity-core/files/13702518/serenity-core-3.2.6-SNAPSHOT.jar.zip)

**Comment by wakaleo** [comment]
The latest snapshot version on master is 4.0.30-SNAPSHOT (https://github.com/serenity-bdd/serenity-core/blob/main/pom.xml) (I don't know where 3.0.26 is coming from, but it would be so old that it would have no hope of working). I have made a change in 4.0.30-SNAPSHOT that might fix things, so that is the version you need to build to try.

**Comment by algernon28** [comment]
I can explain the confusion: from this site I cannot see any 4.x branch, only tagged releases. In fact, this is what I see 
![image](https://github.com/serenity-bdd/serenity-core/assets/11425650/6c1e30de-f64f-442f-9f04-dfcbeb2ab299)
You said "from master branch", so I tried there, and in the "master" (not main) branch all I could find was that 3.2.6-SNAPSHOT.
That said, I managed to build 4.0.30-SNAPSHOT, but it's failing to instantiate the driver. I tried several combinations, could you post the relevant bit of serenity.conf I should use to make it work?
I might be missing something, because:
webdriver.driver=appium
webdriver.remote.driver=appium, and all the combination with remote / appium didn't work (not even reaching sauce, just not failing with instantiating the driver)


[serenity-core-4.0.30-SNAPSHOT.jar.zip](https://github.com/serenity-bdd/serenity-core/files/13714384/serenity-core-4.0.30-SNAPSHOT.jar.zip)

**Comment by wakaleo** [comment]
@algernon28 I don't have an environment to test Appium locally so I don't really know what configuration should be used. The best would be to dig into the code (the driver instantiation code isn't too complex) and see what is going on. In general, `webdriver.driver` will determine what overall driver initialisation logic to use, and `webdriver.remote.driver` is used by the default remote driver logic to determine the default capabilities, and also (in the latest code, as requested) to create an Appium driver instead of a remote driver.

**Comment by algernon28** [comment]
Local Appium works fine, the problem is when going thru Saucelabs.
My serenity.conf actually is like this (i am trying both remote.driver=appium and remote.driver=android): 
I run it with maven clean verify -Denvironment=sauce-android -Dsauce.username=<username> -Dsauce.accessKey=<key>
The capabilities are correct, in fact they work if I run it with a local Appium with a local device/emu.
So two questions:
1) is this serenity.conf correct, ignoring the appium capabilities part?
2) could it be something related to serenity.conf handling? For my local runs I use properties. 

serenity {
    outputDirectory=target/site/serenity
    take.screenshots = AFTER_EACH_STEP
    messages=ui-messages
    output.formats=html
    verbose.steps=true
    reports.show.step.details=true
    timeout=10000
    test{
        root = features
    }
    show.pie.charts = true
    project.name = Edenred Multiwallet Test Automation
}
webdriver{
    timeouts.implicitlywait = 8000
    wait.for.timeout = 60000
}

appium {
    nativeWebScreenshot = true
    autoGrantPermissions=true

}
saucelabs {
  active = true
  username = "${sauce.username}"
  accessKey = "${sauce.accessKey}"
}
environments{
    sauce-android{
        webdriver{
            driver = appium
            remote.driver = android
            remote.url="https://#{saucelabs.username}:#{saucelabs.accessKey}@ondemand.eu-central-1.saucelabs.com:443/wd/hub"
            capabilities{
                "platformName" = "Android"
                "appium:deviceName" = "Google.*"
                "appium:platformVersion" = "13"
                "appium:automationName" = "uiautomator2"
                "appium:app" = "storage:filename=app-uat-release.apk"
                "appium:appPackage" = "com.edenred.mobiletr.staging"
                "appium:appActivity" = "com.edenred.login.ui.activity.LoginActivity"
                "appium:language" = "it"
                "appium:locale" = "IT"
                "sauce:options" {
                    "name" = "Edenred Multiwallet"
                    "appiumVersion" = "latest"
                }
            }
        }

    }
    sauce-ios{
        webdriver{
            driver = "remote"
            remote.url="https://ondemand.eu-central-1.saucelabs.com:443/wd/hub"
            capabilities{
                "appium:platformName" = "iOS"
                "appium:deviceName" = "iPhone.*"
                "appium:platformVersion" = "13"
                "appium:automationName" = "XCUITest"
                "appium:bundleId" = "com.edenred.italyTR.staging"
                "appium:app" = "storage:filename=EdenredMW.ipa"
                "appium:language" = "it"
                "appium:locale" = "IT"
                "sauce:options" {
                    "name" = "Edenred Multiwallet"
                    "appiumVersion" = "latest"

                }
            }
        }
    }
}

**Comment by algernon28** [comment]
John, I suspect the issues also lie in this method from DriverStrategySelector class
    _public SupportedWebDriver forDriverClass(Class<? extends WebDriver> driverClass) {
        return !this.isARemoteDriver(driverClass) && !this.shouldUseARemoteDriver() ? SupportedWebDriver.forClass(driverClass) : SupportedWebDriver.REMOTE;
    }_
.isARemoteDriver(driverClass) returns false, but shouldUseARemoteDriver() returns true, therefore it's switching to REMOTE instead of io.appium.java_client.AppiumDriver, which is the actual driverClass as I look at the debugger.
Basically my understanding is that if there is a webdriver.remote.url, then it should use a remotedriver, therefore it switches to REMOTE.
_ThucydidesSystemProperty.WEBDRIVER_REMOTE_URL.isDefinedIn(this.environmentVariables);_

**Comment by wakaleo** [comment]
Thanks for your analysis. Can you provide a PR with a fix?

**Comment by algernon28** [comment]
That is a partial analysis, meaning it's a problem but I'm not sure if it's the only one. On Sauce side I still see some errors like "isDisplayed is not supported", even if on their appium.log I do see an AndroidUiautomator2Driver instantiated.
Also, in debug it seems it's not resolving most of the ThucydidesSystemProperty  appium properties, which probably means that my serenity.conf should look different.
All in all I haven't solved the problem thru Sauce yet.

**Comment by algernon28** [comment]
Actually I think I need some hints about how the serenity.conf should be structured for Appium and Saucelabs.
Not looking for the individual capabilities, those are documented on Appium website, but I'm a bit at a loss trying to figure out how Serenity expects the capabilities be nested.
As I wrote above, none of the ThucydidesSystemProperty enums seem to apply, or maybe they are only intended for standard properties not sure, but the webdriver instantiation logic does use it.
